dask/array/core.py

```diff
@@ -3493,6 +3498,72 @@ def _vindex(x, *indexes):
     return result_1d.reshape(broadcast_shape + result_1d.shape[1:])
 
 
+def _vindex_dask_arrays(x, *indices):
+    """Point wise indexing with Dask Arrays"""
+
+    from .creation import arange
+
+    x = asarray(x)
+
+    oth_inds = []
+    arr_inds = []
+    out_axes = []
+    for i, ind in enumerate(indices):
+        if isinstance(ind, Array):
+            oth_inds.append(slice(None))
+            arr_inds.append(asarray(ind))
+            if 0 not in out_axes:
+                out_axes.append(0)
+        elif isinstance(ind, Integral):
+            oth_inds.append(ind)
+        else:
+            oth_inds.append(ind)
+            arr_inds.append(slice(None))
+            out_axes.append(len(arr_inds))
+    oth_inds = tuple(oth_inds)
+    arr_inds = tuple(arr_inds)
+    out_axes = tuple(out_axes)
+
+    if not all(o == slice(None) for o in oth_inds):
+        x = x.vindex[oth_inds]
+
+    x_inds = tuple(
+        arange(s, chunks=c) for s, c in zip(x.shape, x.chunks)
+    )
+
+    idx_vals = atop(
+        _vindex_dask_arrays_chunk,
+        out_axes,
+        x, tuple(range(1, 1 + x.ndim)),
+        *concat([
+            (e_0, (0,), e_i, (i,)) if isinstance(e_0, Array) else
+            (e_0, None, e_i, (i,))
+            for i, (e_0, e_i) in enumerate(zip(arr_inds, x_inds), start=1)
+        ]),
+        dtype=x.dtype,
+        concatenate=True
+    )
```

Comment:
I definitely don't understand what's going on here, but that's also pretty typical for non-trivial uses of atop in the codebase, much of which is my fault.  This is a good learning experience for me to be on the other end of this.  I was trying to think of ways to help with this situation.  Perhaps comments for simple cases as in this snippet in `unify_chunks`?

```python
    arginds = [(asarray(a) if ind is not None else a, ind)
               for a, ind in partition(2, args)]  # [x, ij, y, jk]
    args = list(concat(arginds))  # [(x, ij), (y, jk)]
```

Reply:
Yeah, it's a bit ugly. Have been trying to thing of ways to make it clearer. Open to adding comments if that would help.

Basically we pass in our array (probably obvious), `arange` along each dimension in `x`, and then either a 1-D Dask Array that selects points or `slice(None)` (if there is no selection or if it was handled in the first pass). Also we interleaved `arange` and the 1-D selection, which makes this ugly here, but easier to work with in `_vindex_dask_arrays_chunk`. Expect these two last parts makes things confusing.

Reply:
I'm a little concerned about the algorithm here. If I understand it correctly, `atop` here is mapping over all possible pairs of chunks on the indexed and indexing array. This will work okay for small numbers of chunks, but could quickly become intractably slow for larger numbers of chunks.

That said, I can't think of any obvious ways to improve this within dask's data model. You really need to compute some of the indices to know which source chunks they index, but dask (at least the non-distributed version) does not have support for that sort of dynamic graph construction.

Reply:
It may be that it's always better to collapse a chunked index dask-array into a single chunk.  This would increase our work-per-chunk but decrease the scheduling overhead.  In common the case of single-dimensional indexers I suspect that this is almost always a good idea.

Reply:
Tried to add some information about what happens here in a docstring for this function and in a comment above. Suggestions welcome.

Certainly aware that there could be performance problems with large numbers of chunks and would be happy to see them improved. Though it is not unique to this case. That said, it feels like this issue is outside the scope of this PR at present.